### PR TITLE
Add PlayerReleaseItemEvent

### PR DESCRIPTION
**Pull request by u9g**

### CODE REVIEW

**Comment by u9g** [comment]
closes https://github.com/PaperMC/Paper/issues/7766

File: patches/server/0903-Add-PlayerReleaseItemEvent.patch [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+index 1521f53ee1bd85ca44a68b2c9d969eaf63fa342e..4793ae0a49b510b97f1da9d426ddbb214d85ab87 100644
+--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
++++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
+@@ -3868,6 +3868,7 @@ public abstract class LivingEntity extends Entity {
+ 
+     public void releaseUsingItem() {
+         if (!this.useItem.isEmpty()) {
++            new com.destroystokyo.paper.event.player.PlayerReleaseItemEvent((Player)this.getBukkitEntity(), this.useItem.asBukkitCopy(), this.getUseItemRemainingTicks()).callEvent();
```
Comment (Owen1212055) [inline patches/server/0903-Add-PlayerReleaseItemEvent.patch]: Paper comment

Comment (u9g) [reply patches/server/0903-Add-PlayerReleaseItemEvent.patch]: Fixed, Thanks :)

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+Subject: [PATCH] Add PlayerReleaseItemEvent
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerReleaseItemEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerReleaseItemEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..b3b85d54090c567ae05bf68809c47c24ea036e8a
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerReleaseItemEvent.java
+@@ -0,0 +1,52 @@
++package com.destroystokyo.paper.event.player;
```
Comment (Doc94) [inline patches/api/0382-Add-PlayerReleaseItemEvent.patch]: The package need to be io.papermc.paper

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
++    public ItemStack getItem() {
++        return item;
++    }
++
++    /**
++     * Gets the number of ticks the item was held for
++     *
++     * @return int the number of ticks the item was held for
++     */
++    public int getUseItemRemainingTicks() {
```
Comment (electronicboy) [inline patches/api/0382-Add-PlayerReleaseItemEvent.patch]: "item remaing" and the "tickets the item was held for" doesn't really line up

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
++package com.destroystokyo.paper.event.player;
++
++import org.bukkit.entity.Player;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when the server detects a player releasing an item.
```
Comment (electronicboy) [inline patches/api/0382-Add-PlayerReleaseItemEvent.patch]: should probably be better clarified to say that this is "stopped using", idk exactly how to word this but I don't think that this really clarifies what the event is actually covering too well here

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
++ */
++public class PlayerReleaseItemEvent extends PlayerEvent {
```
Comment (electronicboy) [inline patches/api/0382-Add-PlayerReleaseItemEvent.patch]: as above, I kinda eer on the name of this one, I'm not sure it really clarifies what it means

**Review by electronicboy — DISMISSED** [review]

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch (hunk 2 above)
Comment (u9g) [reply patches/api/0382-Add-PlayerReleaseItemEvent.patch]: Fixed!

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch (hunk 3 above)
Comment (u9g) [reply patches/api/0382-Add-PlayerReleaseItemEvent.patch]: renamed to `getTicksHeldFor()`

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch (hunk 4 above)
Comment (u9g) [reply patches/api/0382-Add-PlayerReleaseItemEvent.patch]: renamed event to `PlayerStopUsingItemEvent`

File: patches/api/0382-Add-PlayerReleaseItemEvent.patch (hunk 5 above)
Comment (u9g) [reply patches/api/0382-Add-PlayerReleaseItemEvent.patch]: as said above, renamed the event to `PlayerStopUsingItemEvent` which is much better imo

File: patches/api/0382-Add-PlayerStopUsingItemEvent.patch [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
++        super(player);
++        this.item = item;
++        this.ticksHeldFor = ticksHeldFor;
++    }
++
++    /**
++     * Gets the item the player is releasing
++     *
++     * @return ItemStack the item the player released
++     */
```
Comment (Doc94) [inline patches/api/0382-Add-PlayerStopUsingItemEvent.patch]: If the item passed is a copy i think need to be mentioned here.

**Review by kennytv — APPROVED** [review]

**Pull request merged**